File: packages/android/app/src/main/java/com/viecz/vieczandroid/data/models/TaskApplication.kt
package com.viecz.vieczandroid.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class TaskApplication(
    @Json(name = "id") val id: Long,
    @Json(name = "task_id") val taskId: Long,
    @Json(name = "tasker_id") val taskerId: Long,
    @Json(name = "proposed_price") val proposedPrice: Long? = null,
    @Json(name = "message") val message: String? = null,
    @Json(name = "status") val status: ApplicationStatus,
    @Json(name = "created_at") val createdAt: String,
    @Json(name = "updated_at") val updatedAt: String
)

enum class ApplicationStatus {
    @Json(name = "pending")
    PENDING,
    @Json(name = "accepted")
    ACCEPTED,
    @Json(name = "rejected")
    REJECTED
}

@JsonClass(generateAdapter = true)
data class ApplyTaskRequest(
    @Json(name = "proposed_price") val proposedPrice: Long? = null,
    @Json(name = "message") val message: String? = null
)

@JsonClass(generateAdapter = true)
data class AcceptApplicationResponse(
    @Json(name = "message") val message: String
)

File: server/internal/models/task_application.go
package models

import (
	"fmt"
	"time"

	"gorm.io/gorm"
)

// ApplicationStatus represents the status of a task application
type ApplicationStatus string

const (
	ApplicationStatusPending  ApplicationStatus = "pending"
	ApplicationStatusAccepted ApplicationStatus = "accepted"
	ApplicationStatusRejected ApplicationStatus = "rejected"
)

// TaskApplication represents a tasker's application to a task
type TaskApplication struct {
	ID            int64             `gorm:"primaryKey;autoIncrement" json:"id"`
	TaskID        int64             `gorm:"not null;index" json:"task_id"`
	TaskerID      int64             `gorm:"not null;index" json:"tasker_id"`
	ProposedPrice *int64            `json:"proposed_price,omitempty"`
	Message       *string           `gorm:"size:500" json:"message,omitempty"`
	Status        ApplicationStatus `gorm:"type:varchar(20);default:'pending';index" json:"status"`
	CreatedAt     time.Time         `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt     time.Time         `gorm:"autoUpdateTime" json:"updated_at"`

	// Associations
	Task   Task  `gorm:"foreignKey:TaskID" json:"-"`
	Tasker *User `gorm:"foreignKey:TaskerID" json:"tasker,omitempty"`
}

// Validate validates the task application model
func (ta *TaskApplication) Validate() error {
	if ta.TaskID == 0 {
		return fmt.Errorf("task_id is required")
	}

	if ta.TaskerID == 0 {
		return fmt.Errorf("tasker_id is required")
	}

	if ta.ProposedPrice != nil && *ta.ProposedPrice <= 0 {
		return fmt.Errorf("proposed_price must be greater than 0")
	}

	if ta.Message != nil && len(*ta.Message) > 500 {
		return fmt.Errorf("message must be less than 500 characters")
	}

	validStatuses := map[ApplicationStatus]bool{
		ApplicationStatusPending:  true,
		ApplicationStatusAccepted: true,
		ApplicationStatusRejected: true,
	}
	if !validStatuses[ta.Status] {
		return fmt.Errorf("invalid status: %s", ta.Status)
	}

	return nil
}

// BeforeCreate hook is called before creating a task application
func (ta *TaskApplication) BeforeCreate(tx *gorm.DB) error {
	return ta.Validate()
}

// BeforeUpdate hook is called before updating a task application
func (ta *TaskApplication) BeforeUpdate(tx *gorm.DB) error {
	return ta.Validate()
}
